File: algorithms/fibonacci.py
# implementation of fibonacci series using recursion

def fibonacci(n):
    if n == 0 or n == 1:
        return n
    return fibonacci(n-1) + fibonacci(n-2)

# using dynamic programming
memo = [None] * 100
def fibonacci(n):
    if memo[n] is not None:
        return memo[n]
    if n == 0 or n == 1:
        return n
    memo[n] = fibonacci(n-1) + fibonacci(n-2)
    return memo[n]

# Test cases
print(fibonacci(10)) # 55
print(fibonacci(20)) # 6765

# Time complexity: O(n)
# Space complexity: O(n)
# where n is the number of elements in the fibonacci series
print("")
print("Using dynamic programming")
print(fibonacci(10)) # 55
print(fibonacci(20)) # 6765

File: algorithms/cpp/fibonacci.cpp
// Fibonacci using recursion + dynamic programming (memoization)
// Time complexity:  O(n)
// Space complexity: O(n)

#include <iostream>
#include <unordered_map>

std::unordered_map<int, long long> memo;

long long fibonacci(int n) {
    if (n == 0 || n == 1) return n;
    auto it = memo.find(n);
    if (it != memo.end()) return it->second;
    memo[n] = fibonacci(n - 1) + fibonacci(n - 2);
    return memo[n];
}

int main() {
    std::cout << fibonacci(10) << "\n";  // 55
    std::cout << fibonacci(20) << "\n";  // 6765

    std::cout << "\nUsing dynamic programming\n";
    std::cout << fibonacci(10) << "\n";  // 55
    std::cout << fibonacci(20) << "\n";  // 6765
    return 0;
}
